### PR TITLE
Pass ledger state to consensus

### PR DESCRIPTION
**Pull request by lebdron**
### Description of the Change
Add LedgerState to events from components
Use peer list from LedgerState in YacGateImpl vote

### Benefits
Another step to remove PeerQuery from the pipeline

### Possible Drawbacks
? 

### Usage Examples or Tests
YacPeerOrdererTest

### CODE REVIEW

File: irohad/main/application.cpp [hunk 1]
```diff
@@ -582,11 +582,18 @@ Irohad::RunResult Irohad::run() {
                 std::shared_ptr<shared_model::interface::Block>>>(&block_var)
                              ->value;
 
+            auto peers = storage->createPeerQuery() |
+                [](auto &&peer_query) { return peer_query->getLedgerPeers(); };
```
Comment (igor-egorov) [inline irohad/main/application.cpp:698]: Do we need && here?

File: test/module/irohad/consensus/yac/peer_orderer_test.cpp [hunk 2]
```diff
@@ -42,17 +43,12 @@ class YacPeerOrdererTest : public ::testing::Test {
...
     for (size_t i = 1; i <= N_PEERS; ++i) {
-      auto peer = std::make_shared<MockPeer>();
-      EXPECT_CALL(*peer, address())
-          .WillRepeatedly(ReturnRefOfCopy(std::to_string(i)));
-      EXPECT_CALL(*peer, pubkey())
-          .WillRepeatedly(
-              ReturnRefOfCopy(shared_model::interface::types::PubkeyType(
-                  std::string(32, '0'))));
-
+      auto peer = makePeer(
```
Comment (igor-egorov) [inline test/module/irohad/consensus/yac/peer_orderer_test.cpp:48]: Since you have modified that, could you please rename N_PEERS variable?

kPeers or kPeersQuantity ?

Comment (igor-egorov) [reply test/module/irohad/consensus/yac/peer_orderer_test.cpp:48]: Should not it be at least a constant?

File: irohad/consensus/yac/impl/peer_orderer_impl.cpp [hunk 3]
```diff
@@ -19,23 +19,14 @@ namespace iroha {
...
-              std::default_random_engine gen(seed);
-              std::shuffle(peers.begin(), peers.end(), gen);
-              return ClusterOrdering::create(peers);
-            };
+          const YacHash &hash,
+          std::vector<std::shared_ptr<shared_model::interface::Peer>> peers) {
+        std::seed_seq seed(hash.vote_hashes.block_hash.begin(),
+                           hash.vote_hashes.block_hash.end());
+        std::default_random_engine gen(seed);
+        std::shuffle(peers.begin(), peers.end(), gen);
```
Comment (igor-egorov) [inline irohad/consensus/yac/impl/peer_orderer_impl.cpp:28]: Please reply me, why we use shuffle? What is the end goal?

File: test/module/irohad/consensus/yac/yac_gate_test.cpp [hunk 4]
```diff
@@ -88,6 +88,11 @@ class YacGateTest : public ::testing::Test {
                                          hash_provider,
                                          block_creator,
                                          block_cache);
+
+    auto peer = makePeer("127.0.0.1", shared_model::crypto::PublicKey("111"));
```
Comment (igor-egorov) [inline test/module/irohad/consensus/yac/yac_gate_test.cpp:99]: Should not we use here a public key of proper length?

File: irohad/consensus/gate_object.hpp [hunk 5]
```diff
@@ -11,6 +11,7 @@
 #include "cryptography/hash.hpp"
 #include "cryptography/public_key.hpp"
 #include "interfaces/common_objects/types.hpp"
+#include "ametsuchi/ledger_state.hpp"
```
Comment (Akvinikym) [inline irohad/consensus/gate_object.hpp]: Is that clang-formatted?

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 6]
```diff
@@ -64,7 +67,8 @@ namespace iroha {
                      current_hash_.vote_hashes.block_hash);
         }
 
-        auto order = orderer_->getOrdering(current_hash_);
+        auto order = orderer_->getOrdering(current_hash_,
+                                           *event.ledger_state->ledger_peers);
```
Comment (Akvinikym) [inline irohad/consensus/yac/impl/yac_gate_impl.cpp:76]: Can't the shuffling be done in a more optimal way, for example, via passing not a copy of `std::vector<std::shared_ptr>>`, but `std::vector<int>`, where int-s are indices of peers to be shuffled? Then this order can be passed to `vote`, then those indices can be matched with an original collection, when `vote`-ing

File: irohad/consensus/yac/yac_peer_orderer.hpp [hunk 7]
```diff
@@ -20,20 +20,15 @@ namespace iroha {
...
-         * bot not voted before.
-         * @return ordering, like in ledger
-         */
-        virtual boost::optional<ClusterOrdering> getInitialOrdering() = 0;
-
         /**
          * Provide order of peers based on hash and initial order of peers
          * @param hash - hash-object that used as seed of ordering shuffle
          * @return shuffled cluster order
          */
```
Comment (Akvinikym) [inline irohad/consensus/yac/yac_peer_orderer.hpp:29]: Please, mention `peers` arg as well

File: irohad/main/application.cpp (hunk 1 above)
Comment (Akvinikym) [reply irohad/main/application.cpp:698]: Why not? It is going to accept any kind of reference

File: irohad/simulator/verified_proposal_creator_common.hpp [hunk 8]
```diff
@@ -9,6 +9,7 @@
 #include <boost/optional.hpp>
 #include "consensus/round.hpp"
 #include "validation/stateful_validator_common.hpp"
+#include "ametsuchi/ledger_state.hpp"
```
Comment (Akvinikym) [inline irohad/simulator/verified_proposal_creator_common.hpp]: clang-formatted?

File: test/module/irohad/consensus/yac/peer_orderer_test.cpp [hunk 9]
```diff
@@ -42,17 +43,12 @@ class YacPeerOrdererTest : public ::testing::Test {
     orderer = PeerOrdererImpl(pbfactory);
   }
 
-  std::vector<std::shared_ptr<shared_model::interface::Peer>> peers = [] {
+  std::vector<wPeer> peers = [] {
```
Comment (Akvinikym) [inline test/module/irohad/consensus/yac/peer_orderer_test.cpp]: As I remember, we agreed not to use mentions of pointers kinds in the types

File: test/module/irohad/consensus/yac/yac_gate_test.cpp [hunk 10]
```diff
@@ -88,6 +88,11 @@ class YacGateTest : public ::testing::Test {
...
+    auto ledger_peers =
+        std::make_shared<iroha::PeerList>(iroha::PeerList{peer});
+    ledger_state = std::make_shared<iroha::LedgerState>(ledger_peers);
```
Comment (Akvinikym) [inline test/module/irohad/consensus/yac/yac_gate_test.cpp]:
```suggestion
    ledger_state = std::make_shared<iroha::LedgerState>(std::move(ledger_peers));
```

File: irohad/consensus/yac/impl/yac_gate_impl.cpp (hunk 6 above)
Comment (lebdron) [reply irohad/consensus/yac/impl/yac_gate_impl.cpp:76]: Probably this can be done after profiling if it turns out to be a bottleneck.

File: test/module/irohad/consensus/yac/yac_gate_test.cpp (hunk 4 above)
Comment (lebdron) [reply test/module/irohad/consensus/yac/yac_gate_test.cpp:99]: Probably not, because there is no particular crypto algorithm used in the test, so there is no proper length in this context.

File: irohad/consensus/yac/impl/peer_orderer_impl.cpp (hunk 3 above)
Comment (lebdron) [reply irohad/consensus/yac/impl/peer_orderer_impl.cpp:28]: https://github.com/hyperledger/iroha/blob/536d4043e1d985e8bc21edcd75f19987930f2a9a/irohad/consensus/yac/yac_peer_orderer.hpp#L24
So it is used to generate an order based on hash. Could you suggest a comment to add to the code?

File: irohad/consensus/gate_object.hpp [hunk 11]
```diff
@@ -25,28 +26,33 @@ namespace iroha {
     struct PairValid {
       std::shared_ptr<shared_model::interface::Block> block;
       consensus::Round round;
+      std::shared_ptr<LedgerState> ledger_state;
```
Comment (muratovv) [inline irohad/consensus/gate_object.hpp:27]: I think better to write explicit ctors for any gate_object. If this is unrelated changes, please, leave a todo instead.
